File: app/src/main/java/com/coopra/nebulus/TrackRepository.java
package com.coopra.nebulus;

import android.app.Application;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;
import android.os.AsyncTask;

import com.coopra.database.AppDatabase;
import com.coopra.database.data_access_objects.TrackDao;
import com.coopra.database.data_access_objects.UserDao;
import com.coopra.database.entities.Track;
import com.coopra.database.entities.User;
import com.coopra.nebulus.enums.NetworkStates;

import java.util.List;

public class TrackRepository {
    private TrackDao mTrackDao;
    private UserDao mUserDao;
    private LiveData<PagedList<Track>> mAllTracks;
    private TokenHandler mTokenHandler = new TokenHandler();

    public TrackRepository(Application application, MutableLiveData<NetworkStates> networkState) {
        AppDatabase db = AppDatabase.Companion.getDatabase(application);
        mTrackDao = db.trackDao();
        mUserDao = db.userDao();

        PagedList.Config myPagingConfig = new PagedList.Config.Builder()
                .setPageSize(20)
                .setEnablePlaceholders(false)
                .build();

        mAllTracks = new LivePagedListBuilder<>(mTrackDao.getAll(), myPagingConfig)
                .setBoundaryCallback(new FeedTracksBoundaryCallback(this, mTokenHandler.getToken(application), networkState))
                .build();
    }

    public LiveData<PagedList<Track>> getAll() {
        return mAllTracks;
    }

    public void insertAll(TrackParameters parameters) {
        new insertAllAsyncTask(mTrackDao, mUserDao).execute(parameters);
    }

    private static class insertAllAsyncTask extends AsyncTask<TrackParameters, Void, Void> {
        private TrackDao mAsyncTrackDao;
        private UserDao mAsyncUserDao;

        insertAllAsyncTask(TrackDao trackDao, UserDao userDao) {
            mAsyncTrackDao = trackDao;
            mAsyncUserDao = userDao;
        }

        @Override
        protected Void doInBackground(final TrackParameters... params) {
            mAsyncUserDao.insertAll(params[0].users.toArray(new User[0]));
            mAsyncTrackDao.insertAll(params[0].tracks.toArray(new Track[0]));
            return null;
        }
    }

    public static class TrackParameters {
        List<Track> tracks;
        List<User> users;

        public TrackParameters(List<Track> track, List<User> user) {
            this.tracks = track;
            this.users = user;
        }
    }
}

File: app/src/main/java/com/coopra/nebulus/TrackRepository.kt
package com.coopra.nebulus

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.coopra.database.AppDatabase
import com.coopra.database.data_access_objects.TrackDao
import com.coopra.database.data_access_objects.UserDao
import com.coopra.database.entities.Track
import com.coopra.database.entities.User
import com.coopra.nebulus.enums.NetworkStates

class TrackRepository(application: Application, networkState: MutableLiveData<NetworkStates>) {
    private val trackDao: TrackDao
    private val userDao: UserDao
    private val allTracks: LiveData<PagedList<Track>>

    init {
        val database = AppDatabase.getDatabase(application)
        trackDao = database.trackDao()
        userDao = database.userDao()

        val myPagingConfig = PagedList.Config.Builder()
                .setPageSize(20)
                .setEnablePlaceholders(false)
                .build()

        val tokenHandler = TokenHandler()
        allTracks = LivePagedListBuilder(trackDao.getAll(), myPagingConfig)
                .setBoundaryCallback(FeedTracksBoundaryCallback(this,
                        tokenHandler.getToken(application)!!,
                        networkState))
                .build()
    }

    fun getAll(): LiveData<PagedList<Track>> {
        return allTracks
    }

    suspend fun insertAll(vararg parameters: TrackParameters) {
        userDao.insertAll(*parameters[0].users.toTypedArray())
        trackDao.insertAll(*parameters[0].tracks.toTypedArray())
    }

    data class TrackParameters(val tracks: List<Track>, val users: List<User>)
}